Fails to compile with optparse-applicative-0.17.0.0

```
Criterion/Main/Options.hs:225:7: error:
    • Couldn't match expected type ‘Chunk Doc’
                  with actual type ‘[(Doc, Doc)] -> Chunk Doc’
    • Probable cause: ‘tabulate’ is applied to too few arguments
      In the second argument of ‘($)’, namely
        ‘tabulate [(text n, text d) | (n, (_, d)) <- map f measureKeys]’
      In the expression:
        fmap (text "Regression metrics (for use with --regress):" .$.)
          $ tabulate [(text n, text d) | (n, (_, d)) <- map f measureKeys]
      In an equation for ‘regressionHelp’:
          regressionHelp
            = fmap (text "Regression metrics (for use with --regress):" .$.)
                $ tabulate [(text n, text d) | (n, (_, d)) <- map f measureKeys]
            where
                f k = (k, measureAccessors M.! k)
    |
225 |       tabulate [(text n,text d) | (n,(_,d)) <- map f measureKeys]
    |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Criterion/Main/Options.hs:225:16: error:
    • Couldn't match expected type ‘Int’
                  with actual type ‘[(Doc, Doc)]’
    • In the first argument of ‘tabulate’, namely
        ‘[(text n, text d) | (n, (_, d)) <- map f measureKeys]’
      In the second argument of ‘($)’, namely
        ‘tabulate [(text n, text d) | (n, (_, d)) <- map f measureKeys]’
      In the expression:
        fmap (text "Regression metrics (for use with --regress):" .$.)
          $ tabulate [(text n, text d) | (n, (_, d)) <- map f measureKeys]
    |
225 |       tabulate [(text n,text d) | (n,(_,d)) <- map f measureKeys]
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Please consider making a revision on Hackage.

--- a/.github/workflows/haskell-ci.yml
+++ b/.github/workflows/haskell-ci.yml
@@ -8,9 +8,9 @@
 #
 # For more information, see https://github.com/haskell-CI/haskell-ci
 #
-# version: 0.12
+# version: 0.14
 #
-# REGENDATA ("0.12",["github","--config=cabal.haskell-ci","cabal.project"])
+# REGENDATA ("0.14",["github","--config=cabal.haskell-ci","cabal.project"])
 #
 name: Haskell-CI
 on:
@@ -20,67 +20,131 @@ jobs:
   linux:
     name: Haskell-CI - Linux - ${{ matrix.compiler }}
     runs-on: ubuntu-18.04
+    timeout-minutes:
+      60
     container:
       image: buildpack-deps:bionic
     continue-on-error: ${{ matrix.allow-failure }}
     strategy:
       matrix:
         include:
-          - compiler: ghc-9.0.1
+          - compiler: ghc-9.2.1
+            compilerKind: ghc
+            compilerVersion: 9.2.1
+            setup-method: ghcup
             allow-failure: false
-          - compiler: ghc-8.10.4
+          - compiler: ghc-9.0.2
+            compilerKind: ghc
+            compilerVersion: 9.0.2
+            setup-method: ghcup
+            allow-failure: false
+          - compiler: ghc-8.10.7
+            compilerKind: ghc
+            compilerVersion: 8.10.7
+            setup-method: ghcup
             allow-failure: false
           - compiler: ghc-8.8.4
+            compilerKind: ghc
+            compilerVersion: 8.8.4
+            setup-method: hvr-ppa
             allow-failure: false
           - compiler: ghc-8.6.5
+            compilerKind: ghc
+            compilerVersion: 8.6.5
+            setup-method: hvr-ppa
             allow-failure: false
           - compiler: ghc-8.4.4
+            compilerKind: ghc
+            compilerVersion: 8.4.4
+            setup-method: hvr-ppa
             allow-failure: false
           - compiler: ghc-8.2.2
+            compilerKind: ghc
+            compilerVersion: 8.2.2
+            setup-method: hvr-ppa
             allow-failure: false
           - compiler: ghc-8.0.2
+            compilerKind: ghc
+            compilerVersion: 8.0.2
+            setup-method: hvr-ppa
             allow-failure: false
           - compiler: ghc-7.10.3
+            compilerKind: ghc
+            compilerVersion: 7.10.3
+            setup-method: hvr-ppa
             allow-failure: false
           - compiler: ghc-7.8.4
+            compilerKind: ghc
+            compilerVersion: 7.8.4
+            setup-method: hvr-ppa
             allow-failure: false
           - compiler: ghc-7.6.3
+            compilerKind: ghc
+            compilerVersion: 7.6.3
+            setup-method: hvr-ppa
             allow-failure: false
           - compiler: ghc-7.4.2
+            compilerKind: ghc
+            compilerVersion: 7.4.2
+            setup-method: hvr-ppa
             allow-failure: false
       fail-fast: false
     steps:
       - name: apt
         run: |
           apt-get update
-          apt-get install -y --no-install-recommends gnupg ca-certificates dirmngr curl git software-properties-common
-          apt-add-repository -y 'ppa:hvr/ghc'
-          apt-get update
-          apt-get install -y $CC cabal-install-3.4
+          apt-get install -y --no-install-recommends gnupg ca-certificates dirmngr curl git software-properties-common libtinfo5
+          if [ "${{ matrix.setup-method }}" = ghcup ]; then
+            mkdir -p "$HOME/.ghcup/bin"
+            curl -sL https://downloads.haskell.org/ghcup/0.1.17.3/x86_64-linux-ghcup-0.1.17.3 > "$HOME/.ghcup/bin/ghcup"
+            chmod a+x "$HOME/.ghcup/bin/ghcup"
+            "$HOME/.ghcup/bin/ghcup" install ghc "$HCVER"
+            "$HOME/.ghcup/bin/ghcup" install cabal 3.6.2.0
+          else
+            apt-add-repository -y 'ppa:hvr/ghc'
+            apt-get update
+            apt-get install -y "$HCNAME"
+            mkdir -p "$HOME/.ghcup/bin"
+            curl -sL https://downloads.haskell.org/ghcup/0.1.17.3/x86_64-linux-ghcup-0.1.17.3 > "$HOME/.ghcup/bin/ghcup"
+            chmod a+x "$HOME/.ghcup/bin/ghcup"
+            "$HOME/.ghcup/bin/ghcup" install cabal 3.6.2.0
+          fi
         env:
-          CC: ${{ matrix.compiler }}
+          HCKIND: ${{ matrix.compilerKind }}
+          HCNAME: ${{ matrix.compiler }}
+          HCVER: ${{ matrix.compilerVersion }}
       - name: Set PATH and environment variables
         run: |
           echo "$HOME/.cabal/bin" >> $GITHUB_PATH
-          echo "LANG=C.UTF-8" >> $GITHUB_ENV
-          echo "CABAL_DIR=$HOME/.cabal" >> $GITHUB_ENV
-          echo "CABAL_CONFIG=$HOME/.cabal/config" >> $GITHUB_ENV
-          HCDIR=$(echo "/opt/$CC" | sed 's/-/\//')
-          HCNAME=ghc
-          HC=$HCDIR/bin/$HCNAME
-          echo "HC=$HC" >> $GITHUB_ENV
-          echo "HCPKG=$HCDIR/bin/$HCNAME-pkg" >> $GITHUB_ENV
-          echo "HADDOCK=$HCDIR/bin/haddock" >> $GITHUB_ENV
-          echo "CABAL=/opt/cabal/3.4/bin/cabal -vnormal+nowrap" >> $GITHUB_ENV
+          echo "LANG=C.UTF-8" >> "$GITHUB_ENV"
+          echo "CABAL_DIR=$HOME/.cabal" >> "$GITHUB_ENV"
+          echo "CABAL_CONFIG=$HOME/.cabal/config" >> "$GITHUB_ENV"
+          HCDIR=/opt/$HCKIND/$HCVER
+          if [ "${{ matrix.setup-method }}" = ghcup ]; then
+            HC=$HOME/.ghcup/bin/$HCKIND-$HCVER
+            echo "HC=$HC" >> "$GITHUB_ENV"
+            echo "HCPKG=$HOME/.ghcup/bin/$HCKIND-pkg-$HCVER" >> "$GITHUB_ENV"
+            echo "HADDOCK=$HOME/.ghcup/bin/haddock-$HCVER" >> "$GITHUB_ENV"
+            echo "CABAL=$HOME/.ghcup/bin/cabal-3.6.2.0 -vnormal+nowrap" >> "$GITHUB_ENV"
+          else
+            HC=$HCDIR/bin/$HCKIND
+            echo "HC=$HC" >> "$GITHUB_ENV"
+            echo "HCPKG=$HCDIR/bin/$HCKIND-pkg" >> "$GITHUB_ENV"
+            echo "HADDOCK=$HCDIR/bin/haddock" >> "$GITHUB_ENV"
+            echo "CABAL=$HOME/.ghcup/bin/cabal-3.6.2.0 -vnormal+nowrap" >> "$GITHUB_ENV"
+          fi
+
           HCNUMVER=$(${HC} --numeric-version|perl -ne '/^(\d+)\.(\d+)\.(\d+)(\.(\d+))?$/; print(10000 * $1 + 100 * $2 + ($3 == 0 ? $5 != 1 : $3))')
-          echo "HCNUMVER=$HCNUMVER" >> $GITHUB_ENV
-          echo "ARG_TESTS=--enable-tests" >> $GITHUB_ENV
-          echo "ARG_BENCH=--enable-benchmarks" >> $GITHUB_ENV
-          echo "HEADHACKAGE=false" >> $GITHUB_ENV
-          echo "ARG_COMPILER=--$HCNAME --with-compiler=$HC" >> $GITHUB_ENV
-          echo "GHCJSARITH=0" >> $GITHUB_ENV
+          echo "HCNUMVER=$HCNUMVER" >> "$GITHUB_ENV"
+          echo "ARG_TESTS=--enable-tests" >> "$GITHUB_ENV"
+          echo "ARG_BENCH=--enable-benchmarks" >> "$GITHUB_ENV"
+          echo "HEADHACKAGE=false" >> "$GITHUB_ENV"
+          echo "ARG_COMPILER=--$HCKIND --with-compiler=$HC" >> "$GITHUB_ENV"
+          echo "GHCJSARITH=0" >> "$GITHUB_ENV"
         env:
-          CC: ${{ matrix.compiler }}
+          HCKIND: ${{ matrix.compilerKind }}
+          HCNAME: ${{ matrix.compiler }}
+          HCVER: ${{ matrix.compilerVersion }}
       - name: env
         run: |
           env
@@ -102,6 +166,10 @@ jobs:
             prefix: $CABAL_DIR
           repository hackage.haskell.org
             url: http://hackage.haskell.org/
+          EOF
+          cat >> $CABAL_CONFIG <<EOF
+          program-default-options
+            ghc-options: $GHCJOBS +RTS -M3G -RTS
           EOF
           cat $CABAL_CONFIG
       - name: versions
@@ -143,11 +211,12 @@ jobs:
       - name: generate cabal.project
         run: |
           PKGDIR_criterion="$(find "$GITHUB_WORKSPACE/unpacked" -maxdepth 1 -type d -regex '.*/criterion-[0-9.]*')"
-          echo "PKGDIR_criterion=${PKGDIR_criterion}" >> $GITHUB_ENV
+          echo "PKGDIR_criterion=${PKGDIR_criterion}" >> "$GITHUB_ENV"
           PKGDIR_criterion_measurement="$(find "$GITHUB_WORKSPACE/unpacked" -maxdepth 1 -type d -regex '.*/criterion-measurement-[0-9.]*')"
-          echo "PKGDIR_criterion_measurement=${PKGDIR_criterion_measurement}" >> $GITHUB_ENV
+          echo "PKGDIR_criterion_measurement=${PKGDIR_criterion_measurement}" >> "$GITHUB_ENV"
           PKGDIR_criterion_examples="$(find "$GITHUB_WORKSPACE/unpacked" -maxdepth 1 -type d -regex '.*/criterion-examples-[0-9.]*')"
-          echo "PKGDIR_criterion_examples=${PKGDIR_criterion_examples}" >> $GITHUB_ENV
+          echo "PKGDIR_criterion_examples=${PKGDIR_criterion_examples}" >> "$GITHUB_ENV"
+          rm -f cabal.project cabal.project.local
           touch cabal.project
           touch cabal.project.local
           echo "packages: ${PKGDIR_criterion}" >> cabal.project
@@ -200,10 +269,12 @@ jobs:
           rm -f cabal.project.local
       - name: constraint set fast
         run: |
+          $CABAL v2-build $ARG_COMPILER --enable-tests --disable-benchmarks --constraint='criterion-measurement +fast' --constraint='criterion +fast' --dependencies-only -j2 all
           $CABAL v2-build $ARG_COMPILER --enable-tests --disable-benchmarks --constraint='criterion-measurement +fast' --constraint='criterion +fast' all
           $CABAL v2-test $ARG_COMPILER --enable-tests --disable-benchmarks --constraint='criterion-measurement +fast' --constraint='criterion +fast' all
       - name: constraint set embed-data-files
         run: |
+          $CABAL v2-build $ARG_COMPILER --enable-tests --enable-benchmarks --constraint='criterion +embed-data-files' --dependencies-only -j2 all
           $CABAL v2-build $ARG_COMPILER --enable-tests --enable-benchmarks --constraint='criterion +embed-data-files' all
           $CABAL v2-test $ARG_COMPILER --enable-tests --enable-benchmarks --constraint='criterion +embed-data-files' all
           $CABAL v2-haddock $ARG_COMPILER --with-haddock $HADDOCK --enable-tests --enable-benchmarks --constraint='criterion +embed-data-files' all

--- a/Criterion/Main/Options.hs
+++ b/Criterion/Main/Options.hs
@@ -25,6 +25,7 @@ module Criterion.Main.Options
 
 import Control.Monad (when)
 import Criterion.Analysis (validateAccessors)
+import Criterion.Main.Options.Internal (tabulate)
 import Criterion.Types (Config(..), Verbosity(..), measureAccessors,
                         measureKeys)
 import Data.Char (isSpace, toLower)
@@ -34,7 +35,7 @@ import Data.List (isPrefixOf)
 import Data.Version (showVersion)
 import GHC.Generics (Generic)
 import Options.Applicative
-import Options.Applicative.Help (Chunk(..), tabulate)
+import Options.Applicative.Help (Chunk(..))
 import Options.Applicative.Help.Pretty ((.$.))
 import Options.Applicative.Types
 import Paths_criterion (version)

--- a/Criterion/Main/Options/Internal.hs
+++ b/Criterion/Main/Options/Internal.hs
@@ -1,0 +1,38 @@
+{-# LANGUAGE CPP #-}
+
+-- |
+-- Module      : Criterion.Main.Options.Internal
+-- Copyright   : (c) 2022 Ryan Scott
+--
+-- License     : BSD-style
+-- Maintainer  : bos@serpentine.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Provides a shim on top of @optparse-applicative@'s 'Options.tabulate'
+-- function that is backwards-compatible with pre-@0.17.*@ versions of
+-- @optparse-applicative@. This is deliberately kept separate from the rest of
+-- "Criterion.Main.Options" because this function requires CPP to define, and
+-- there is a Haddock comment in "Criterion.Main.Options" that will cause the
+-- CPP preprocessor to trigger an \"unterminated comment\" error. Ugh.
+--
+-- TODO: When we support @optparse-applicative-0.17@ as the minimum, remove
+-- this module and simply inline the definition of 'tabulate' in
+-- "Criterion.Main.Options".
+module Criterion.Main.Options.Internal (tabulate) where
+
+import qualified Options.Applicative.Help as Options
+import Options.Applicative.Help (Chunk, Doc)
+
+#if MIN_VERSION_optparse_applicative(0,17,0)
+import Options.Applicative (ParserPrefs(..), defaultPrefs)
+#endif
+
+-- | A shim on top of 'Options.tabulate' from @optparse-applicative@ that is
+-- backwards-compatible with pre-@0.17.*@ versions of @optparse-applicative@.
+tabulate :: [(Doc, Doc)] -> Chunk Doc
+#if MIN_VERSION_optparse_applicative(0,17,0)
+tabulate = Options.tabulate (prefTabulateFill defaultPrefs)
+#else
+tabulate = Options.tabulate
+#endif

--- a/changelog.md
+++ b/changelog.md
@@ -1,3 +1,7 @@
+next
+
+* Allow building with `optparse-applicative-0.17.*`.
+
 1.5.12.0
 
 * Fix a bug introduced in version 1.5.9.0 in which benchmark names that include

--- a/criterion-measurement/criterion-measurement.cabal
+++ b/criterion-measurement/criterion-measurement.cabal
@@ -22,8 +22,9 @@ tested-with:
   GHC==8.4.4,
   GHC==8.6.5,
   GHC==8.8.4,
-  GHC==8.10.4,
-  GHC==9.0.1
+  GHC==8.10.7,
+  GHC==9.0.2,
+  GHC==9.2.1
 
 flag fast
   description: compile without optimizations

--- a/criterion.cabal
+++ b/criterion.cabal
@@ -74,6 +74,7 @@ library
     Criterion.Types
 
   other-modules:
+    Criterion.Main.Options.Internal
     Criterion.Monad.Internal
 
   other-modules:
@@ -102,7 +103,8 @@ library
     js-chart >= 2.9.4 && < 3,
     mtl >= 2,
     mwc-random >= 0.8.0.3,
-    optparse-applicative >= 0.13,
+    -- TODO: Depend on optparse-applicative-0.17 as the minimum (see #258)
+    optparse-applicative >= 0.13 && < 0.18,
     parsec >= 3.1.0,
     statistics >= 0.14 && < 0.16,
     text >= 0.11,

--- a/criterion.cabal
+++ b/criterion.cabal
@@ -27,8 +27,9 @@ tested-with:
   GHC==8.4.4,
   GHC==8.6.5,
   GHC==8.8.4,
-  GHC==8.10.4,
-  GHC==9.0.1
+  GHC==8.10.7,
+  GHC==9.0.2,
+  GHC==9.2.1
 
 data-files:
   templates/*.css

--- a/examples/criterion-examples.cabal
+++ b/examples/criterion-examples.cabal
@@ -20,8 +20,9 @@ tested-with:
   GHC==8.4.4,
   GHC==8.6.5,
   GHC==8.8.4,
-  GHC==8.10.4,
-  GHC==9.0.1
+  GHC==8.10.7,
+  GHC==9.0.2,
+  GHC==9.2.1
 
 flag conduit-vs-pipes
   default: True

